src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/24.diff.sql

```diff
@@ -0,0 +1,1176 @@
+GO
+
+-- Deleting duplicate rows based on all columns
+EXEC dbo.LogSchemaMigrationProgress 'Deleting redundant rows from dbo.StringSearchParam'
+GO
+WITH cte AS (
+    SELECT ResourceTypeId, ResourceSurrogateId, SearchParamId, Text, TextOverflow, IsMin, IsMax, IsHistory, ROW_NUMBER() 
+    OVER (
+		PARTITION BY ResourceTypeId, ResourceSurrogateId, SearchParamId, Text, TextOverflow, IsMin, IsMax, IsHistory
+		ORDER BY ResourceTypeId, ResourceSurrogateId, SearchParamId, Text
+	) row_num
+	FROM dbo.StringSearchParam
+)
+DELETE FROM cte WHERE row_num > 1
+GO
+
+EXEC dbo.LogSchemaMigrationProgress 'Adding BulkStringSearchParamTableType_3'
+IF TYPE_ID(N'BulkStringSearchParamTableType_3') IS NULL
+BEGIN
+    CREATE TYPE dbo.BulkStringSearchParamTableType_3 AS TABLE
+    (
+        Offset int NOT NULL,
+        SearchParamId smallint NOT NULL,
+        Text nvarchar(256) COLLATE Latin1_General_100_CI_AI_SC NOT NULL,
+        TextOverflow nvarchar(max) COLLATE Latin1_General_100_CI_AI_SC NULL,
+        IsMin bit NOT NULL,
+        IsMax bit NOT NULL,
+        TextHash nvarchar(32) COLLATE Latin1_General_100_CI_AI_SC NOT NULL
+    )
+END
+GO
+
+/*************************************************************
+Insert TextHash column and backfill for existing rows
+**************************************************************/
+EXEC dbo.LogSchemaMigrationProgress 'Adding TextHash column in dbo.StringSearchParam'
+IF NOT EXISTS (
+    SELECT * 
+    FROM   sys.columns 
+    WHERE  object_id = OBJECT_ID('dbo.StringSearchParam') AND name = 'TextHash')
+BEGIN
+    EXEC dbo.LogSchemaMigrationProgress 'Adding TextHash column as nullable';
+    ALTER TABLE dbo.StringSearchParam 
+        ADD TextHash nvarchar(32) NULL 
+END
+GO
+
+-- Backfill values for TextHash column and update it as NOT NULL
+IF EXISTS (
+    SELECT * 
+    FROM   sys.columns 
+    WHERE  object_id = OBJECT_ID('dbo.StringSearchParam') AND name = 'TextHash' AND is_nullable = 1)
+BEGIN
+    EXEC dbo.LogSchemaMigrationProgress 'Backfilling values to TextHash column';
+    UPDATE dbo.StringSearchParam 
+        SET TextHash = (CONVERT([nvarchar](32), (hashbytes('SHA2_256', CASE
+							WHEN [TextOverflow] IS NOT NULL
+							THEN [TextOverflow]
+							ELSE [Text]
+							END)),2))
+
+    EXEC dbo.LogSchemaMigrationProgress 'Update TextHash as NOT NULL';
+    ALTER TABLE dbo.StringSearchParam ALTER COLUMN TextHash nvarchar(32) NOT NULL
+END
+GO
+
+/*************************************************************
+Add Primary key for dbo.StringSearchParam
+**************************************************************/
+IF NOT EXISTS (
+    SELECT * 
+	FROM sys.key_constraints 
+	WHERE name='PK_StringSearchParam' AND type='PK')
+BEGIN
+    EXEC dbo.LogSchemaMigrationProgress 'Adding PK_StringSearchParam'
+	ALTER TABLE dbo.StringSearchParam 
+	ADD CONSTRAINT PK_StringSearchParam PRIMARY KEY NONCLUSTERED (ResourceTypeId, ResourceSurrogateId, SearchParamId, TextHash)
```

Comment:
>PK_StringSearchParam [](http://example.com/codeflow?start=16&length=20)

let consider following example.

I have Patient record and that patient have two duplicate names in it.
```
"name": [
    {
      "use": "official",
      "family": "Chalmers",
      "given": [
        "Peter",
        "James"
      ]
    },
     {
      "use": "official",
      "family": "Chalmers",
      "given": [
        "Peter",
        "James"
      ]
    },

```

Is it valid FHIR resource? I think it is.

This is what I would have in our db if I insert it :

ResourceTypeId ResourceSurrogateId SearchParamId Text
103 5102022608830401891 1016 Chalmers
103 5102022608830401891 1016 Chalmers


A duplicate!

This records gonna have same hash, and therefore, can be inserted.

So first of all, I don't think script handles such case during conversion, and would just fail.

Second, even if script would work, I think it app throw me an error during insertion. On a valid resource, btw.
No one would like it.

Reply:
Thanks for the catch!
So I look into this and figure out that isMin(or isMax) values are set differently though the primary data is similar. 
![image](https://user-images.githubusercontent.com/57157506/146469378-371649c1-8d4c-41dc-bff2-d3c76edc13d3.png)

There are two ways to approach this-
1. Include isMin and isMax in the primary key combination
        But it seems we are consider one of the two rows in our searches so the other row would just be eating space
2. Update application code to have same or consistent values for isMin and isMax when we get duplicate objects for search indices and then the respective rowGenerator class, would only insert distinct row into the table.
        This would need additional work and we would need to consider import operation too.

Hybrid approach-
Have isMin and isMax in primary key since they are just bits and this would ensure no errors while insert but we can work as a separate item to update our application code to have consistent values for isMin and isMax.
Thoughts?

Reply:
1. IsMIn and IsMax. You can have only 4 combinations of them, FHIR spec doesn't stop me from having as many duplicate name records as I want to.

2. Maybe it would work.